In [1]:
import pandas as pd

# Limpeza dos datasets

A limpeza de cada dataset segue os [seguintes passos](https://github.com/Malware-Hunter/Data_Preparation):

1. Remoção de valores faltantes
2. Remoção de amostras com valores duplicados nas colunas
3. Remoção de características irrelevantes (colunas com apenas um valor em todas as linhas)
4. Conversão de tipos de dados

In [2]:
## Funções úteis

def get_unique_values(df):
    for column_name in df.columns:
        yield (column_name, df[column_name].unique())

def drop_irrelevant_columns(df):
    # retorna o df sem colunas irrelevantes (colunas com menos de 2 valores possíveis)
    irrelevant_columns = []
    for (column_name, unique_values) in get_unique_values(df):
        if(len(unique_values) < 2):
            irrelevant_columns.append(column_name)
    print(f"{len(irrelevant_columns)} colunas irrelevantes:")
    print(irrelevant_columns)
    return df.drop(columns=irrelevant_columns)

def cast_to_numeric_inplace(df):
    for non_numeric_column in df.select_dtypes(exclude=['int','float']).columns:
        df[non_numeric_column] = pd.to_numeric(df[non_numeric_column])

## Androcrawl Dataset

In [3]:
df = pd.read_csv('./androcrawl.csv', low_memory=False)
df.columns = [name.replace("'", "") for name in df.columns]
df = df.rename(columns={'Detection Ratio': 'class'})
df

Apk Name  \
0       51113d15a58841b8c81bc7fc8fd7b45f4ca19a29029b86...   
1       3f944d58fbb88b8c64c82c3cd60b12d851176df8aca0d2...   
2       1370ea8b3e2d895b83b5a7771348546eb397624cdc52aa...   
3       e0ac1ffb30426d6f650fa7b3244277f273220fbe230ef6...   
4       0ede82b2a99ed040e65ce1816f054921187d184320301f...   
...                                                   ...   
162978  559f0610501a1bd0f3e7265494cec8537bb0d8c2da47bb...   
162979  a4f11ba03162b949aa29221d5bb2e4909bb116e8cc41a9...   
162980  1f3d156bf6ac206e4cf79db74065865eaf414f181b3bcd...   
162981  cc4b6e06f05a34779803e5f0d25b226a0a9cd10ba5453c...   
162982  eda85939db98c9f5283ce9304f40ea1f841f25a8586b13...   

        Dangerous Permission: ACCESS_SUPERUSER  \
0                                        False   
1                                        False   
2                                        False   
3                                        False   
4                                        False   
...                                        ...   
162978                                   False   
162979                                   False   
162980                                   False   
162981                                   False   
162982                                   False   

        Dangerous Permission: BLUETOOTH_PRIVILEGED  \
0                                            False   
1                                            False   
2                                            False   
3                                            False   
4                                            False   
...                                            ...   
162978                                       False   
162979                                       False   
162980                                       False   
162981                                       False   
162982                                       False   

        Dangerous Permission: BRICK  \
0                             False   
1                             False   
2                             False   
3                             False   
4                             False   
...                             ...   
162978                        False   
162979                        False   
162980                        False   
162981                        False   
162982                        False   

        Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE  \
0                                                   False      
1                                                   False      
2                                                   False      
3                                                   False      
4                                                   False      
...                                                   ...      
162978                                              False      
162979                                              False      
162980                                              False      
162981                                              False      
162982                                              False      

        Dangerous Permission: CLEAR_APP_USER_DATA  \
0                                           False   
1                                           False   
2                                           False   
3                                           False   
4                                           False   
...                                           ...   
162978                                      False   
162979                                      False   
162980                                      False   
162981                                      False   
162982                                      False   

        Dangerous Permission: DELETE_CACHE_FILES  \
0                                          False   
1                                          False   
2                                  

### 1) Remoção de valores faltantes

In [93]:
df = df.dropna()
df

Apk Name  \
0       51113d15a58841b8c81bc7fc8fd7b45f4ca19a29029b86...   
1       3f944d58fbb88b8c64c82c3cd60b12d851176df8aca0d2...   
2       1370ea8b3e2d895b83b5a7771348546eb397624cdc52aa...   
3       e0ac1ffb30426d6f650fa7b3244277f273220fbe230ef6...   
4       0ede82b2a99ed040e65ce1816f054921187d184320301f...   
...                                                   ...   
162978  559f0610501a1bd0f3e7265494cec8537bb0d8c2da47bb...   
162979  a4f11ba03162b949aa29221d5bb2e4909bb116e8cc41a9...   
162980  1f3d156bf6ac206e4cf79db74065865eaf414f181b3bcd...   
162981  cc4b6e06f05a34779803e5f0d25b226a0a9cd10ba5453c...   
162982  eda85939db98c9f5283ce9304f40ea1f841f25a8586b13...   

        Dangerous Permission: ACCESS_SUPERUSER  \
0                                        False   
1                                        False   
2                                        False   
3                                        False   
4                                        False   
...                                        ...   
162978                                   False   
162979                                   False   
162980                                   False   
162981                                   False   
162982                                   False   

        Dangerous Permission: BLUETOOTH_PRIVILEGED  \
0                                            False   
1                                            False   
2                                            False   
3                                            False   
4                                            False   
...                                            ...   
162978                                       False   
162979                                       False   
162980                                       False   
162981                                       False   
162982                                       False   

        Dangerous Permission: BRICK  \
0                             False   
1                             False   
2                             False   
3                             False   
4                             False   
...                             ...   
162978                        False   
162979                        False   
162980                        False   
162981                        False   
162982                        False   

        Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE  \
0                                                   False      
1                                                   False      
2                                                   False      
3                                                   False      
4                                                   False      
...                                                   ...      
162978                                              False      
162979                                              False      
162980                                              False      
162981                                              False      
162982                                              False      

        Dangerous Permission: CLEAR_APP_USER_DATA  \
0                                           False   
1                                           False   
2                                           False   
3                                           False   
4                                           False   
...                                           ...   
162978                                      False   
162979                                      False   
162980                                      False   
162981                                      False   
162982                                      False   

        Dangerous Permission: DELETE_CACHE_FILES  \
0                                          False   
1                                          False   
2                                  

In [94]:
print("Valores únicos para cada feature:")
for (column_name, unique_values) in get_unique_values(df):
    print(column_name, unique_values)

Valores únicos para cada feature:
Apk Name ['51113d15a58841b8c81bc7fc8fd7b45f4ca19a29029b86e4714f1451aa078340'
 '3f944d58fbb88b8c64c82c3cd60b12d851176df8aca0d2438013f654be474f72'
 '1370ea8b3e2d895b83b5a7771348546eb397624cdc52aae88c1e63fbcd0e37ee' ...
 'a4f11ba03162b949aa29221d5bb2e4909bb116e8cc41a9087a27c7c9dfd76eb9'
 'cc4b6e06f05a34779803e5f0d25b226a0a9cd10ba5453c8ad90e3995584fe245'
 'eda85939db98c9f5283ce9304f40ea1f841f25a8586b13e4a9024befc47e3970']
Dangerous Permission: ACCESS_SUPERUSER [False  True]
Dangerous Permission: BLUETOOTH_PRIVILEGED [False]
Dangerous Permission: BRICK [False]
Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE [False  True]
Dangerous Permission: CLEAR_APP_USER_DATA [False  True]
Dangerous Permission: DELETE_CACHE_FILES [False  True]
Dangerous Permission: DELETE_PACKAGES [False  True]
Dangerous Permission: DISABLE_KEYGUARD [False  True]
Dangerous Permission: FACTORY_TEST [False  True]
Dangerous Permission: INSTALL_PACKAGES [False  True]
Dangerous Permissio

Calls a System Routine: ls ['false' '?']
Calls a System Routine: loadjar ['false' '?']
Calls a System Routine: grep ['false' '?']
Calls a System Routine: /sh ['false' '?']
Calls a System Routine: /bin ['false' '?' 'true']
Calls a System Routine: pm install ['false' '?']
Calls a System Routine: /dev/net ['false' '?']
Calls a System Routine: insmod ['false' '?']
Calls a System Routine: rm ['false' '?']
Calls a System Routine: mount ['false' '?' 'true']
Calls a System Routine: root ['false' '?']
Calls a System Routine: /system ['false' '?' 'true']
Calls a System Routine: stdout ['false' '?']
Calls a System Routine: reboot ['false' '?']
Calls a System Routine: killall ['false' '?']
Calls a System Routine: chmod ['false' '?']
Calls a System Routine: stderr ['false' '?']
Calls a System Routine: ratc ['false' '?']
Harmless Permission: ACCESS_SURFACE_FLINGER [False  True]
Harmless Permission: ACCOUNT_MANAGER [False  True]
Harmless Permission: ADD_VOICEMAIL [False  True]
Harmless Permission: CO

In [95]:
count = 0
for (column_name, unique_values) in get_unique_values(df):
    if('?' in unique_values):
        count += 1
print("# de colunas com o valor de interrogação:", count)

<ipython-input-95-6ee8f078b5a6>:3: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  if('?' in unique_values):


# de colunas com o valor de interrogação: 159


Nota-se que o valor '?' ocorre na maioria das colunas (159 de um total de 222), então vamos remover apenas as amostras em que esse valor aparece. O número de colunas já será reduzido após a remoção de colunas irrelevantes no passo 3.

In [96]:
df = df[~df.isin(['?']).any(axis=1)]
df

Apk Name  \
0       51113d15a58841b8c81bc7fc8fd7b45f4ca19a29029b86...   
2       1370ea8b3e2d895b83b5a7771348546eb397624cdc52aa...   
3       e0ac1ffb30426d6f650fa7b3244277f273220fbe230ef6...   
4       0ede82b2a99ed040e65ce1816f054921187d184320301f...   
5       822967e368d00b993e90c0724bed3daf26ba9b59f3a55e...   
...                                                   ...   
162978  559f0610501a1bd0f3e7265494cec8537bb0d8c2da47bb...   
162979  a4f11ba03162b949aa29221d5bb2e4909bb116e8cc41a9...   
162980  1f3d156bf6ac206e4cf79db74065865eaf414f181b3bcd...   
162981  cc4b6e06f05a34779803e5f0d25b226a0a9cd10ba5453c...   
162982  eda85939db98c9f5283ce9304f40ea1f841f25a8586b13...   

        Dangerous Permission: ACCESS_SUPERUSER  \
0                                        False   
2                                        False   
3                                        False   
4                                        False   
5                                        False   
...                                        ...   
162978                                   False   
162979                                   False   
162980                                   False   
162981                                   False   
162982                                   False   

        Dangerous Permission: BLUETOOTH_PRIVILEGED  \
0                                            False   
2                                            False   
3                                            False   
4                                            False   
5                                            False   
...                                            ...   
162978                                       False   
162979                                       False   
162980                                       False   
162981                                       False   
162982                                       False   

        Dangerous Permission: BRICK  \
0                             False   
2                             False   
3                             False   
4                             False   
5                             False   
...                             ...   
162978                        False   
162979                        False   
162980                        False   
162981                        False   
162982                        False   

        Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE  \
0                                                   False      
2                                                   False      
3                                                   False      
4                                                   False      
5                                                   False      
...                                                   ...      
162978                                              False      
162979                                              False      
162980                                              False      
162981                                              False      
162982                                              False      

        Dangerous Permission: CLEAR_APP_USER_DATA  \
0                                           False   
2                                           False   
3                                           False   
4                                           False   
5                                           False   
...                                           ...   
162978                                      False   
162979                                      False   
162980                                      False   
162981                                      False   
162982                                      False   

        Dangerous Permission: DELETE_CACHE_FILES  \
0                                          False   
2                                          False   
3                                  

### 3) Remoção de características irrelevantes

In [97]:
df = drop_irrelevant_columns(df)
df = df.drop(columns=['Apk Name'])
df

67 colunas irrelevantes:
['Dangerous Permission: BLUETOOTH_PRIVILEGED', 'Dangerous Permission: BRICK', 'Dangerous Permission: MOUNT_UNMOUNT_FILESYSTEM', 'Api Call: Landroid/telephony/cdma/CdmaCellLocation;->getSystemId', 'Api Call: Landroid/content/pm/PackageManager;->queryContentProviders', 'Api Call: Landroid/content/BroadcastReceiver;->abortBroadcast', 'Checks adb_enabled', 'Tries to modify adb_enabled', 'Can Steal Data: CAPTURE_AUDIO_OUTPUT', 'Can Steal Data: CAPTURE_SECURE_VIDEO_OUTPUT', 'Can Steal Data: CAPTURE_VIDEO_OUTPUT', 'Can Steal Data: LOCATION_HARDWARE', 'Can Steal Data: READ_SOCIAL_STREAM', 'Content: content://com.android.calendar', 'Content: content://calendar', 'Content: content://com.facebook.katana.provider.AttributionIdProvider', 'Content: content://media', 'C2M Permission: c2dm.permission.RECEIVE', 'C2M Permission: C2D_MESSAGE', 'C2M Permission: c2dm.permission.SEND', 'C2M Intent: c2dm.intent.RECEIVE', 'C2M Intent: c2dm.intent.REGISTRATION', 'C2M Intent: c2dm.inten

Dangerous Permission: ACCESS_SUPERUSER  \
0                                        False   
2                                        False   
3                                        False   
4                                        False   
5                                        False   
...                                        ...   
162978                                   False   
162979                                   False   
162980                                   False   
162981                                   False   
162982                                   False   

        Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE  \
0                                                   False      
2                                                   False      
3                                                   False      
4                                                   False      
5                                                   False      
...                                                   ...      
162978                                              False      
162979                                              False      
162980                                              False      
162981                                              False      
162982                                              False      

        Dangerous Permission: CLEAR_APP_USER_DATA  \
0                                           False   
2                                           False   
3                                           False   
4                                           False   
5                                           False   
...                                           ...   
162978                                      False   
162979                                      False   
162980                                      False   
162981                                      False   
162982                                      False   

        Dangerous Permission: DELETE_CACHE_FILES  \
0                                          False   
2                                          False   
3                                          False   
4                                          False   
5                                          False   
...                                          ...   
162978                                     False   
162979                                     False   
162980                                     False   
162981                                     False   
162982                                     False   

        Dangerous Permission: DELETE_PACKAGES  \
0                                       False   
2                                       False   
3                                       False   
4                                       False   
5                                       False   
...                                       ...   
162978                                  False   
162979                                  False   
162980                                  False   
162981                                  False   
162982                                  False   

        Dangerous Permission: DISABLE_KEYGUARD  \
0                                        False   
2                                        False   
3                                        False   
4                                        False   
5                                        False   
...                                        ...   
162978                                   False   
162979                                   False   
162980                                   False   
162981                                   False   
162982                                   False   

        Dangerous Permission: FACTORY_TEST  \
0                                    False   
2                                    False   
3                                    False   
4 

### 4) Conversão de tipos de dados

In [98]:
df['class'] = df['class'].replace({'Benign' : 0, 'Malicious': 1})
df = df.replace(['false', 'true', False, True], [0, 1, 0, 1])
df = df.drop_duplicates()
df.head()

Dangerous Permission: ACCESS_SUPERUSER  \
0                                       0   
2                                       0   
3                                       0   
4                                       0   
5                                       0   

   Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE  \
0                                                  0      
2                                                  0      
3                                                  0      
4                                                  0      
5                                                  0      

   Dangerous Permission: CLEAR_APP_USER_DATA  \
0                                          0   
2                                          0   
3                                          0   
4                                          0   
5                                          0   

   Dangerous Permission: DELETE_CACHE_FILES  \
0                                         0   
2                                         0   
3                                         0   
4                                         0   
5                                         0   

   Dangerous Permission: DELETE_PACKAGES  \
0                                      0   
2                                      0   
3                                      0   
4                                      0   
5                                      0   

   Dangerous Permission: DISABLE_KEYGUARD  Dangerous Permission: FACTORY_TEST  \
0                                       0                                   0   
2                                       0                                   0   
3                                       0                                   0   
4                                       0                                   0   
5                                       0                                   0   

   Dangerous Permission: INSTALL_PACKAGES  \
0                                       0   
2                                       0   
3                                       0   
4                                       0   
5                                       0   

   Dangerous Permission: INJECT_EVENTS  \
0                                    0   
2                                    0   
3                                    0   
4                                    0   
5                                    0   

   Dangerous Permission: INTERNAL_SYSTEM_WINDOW  ...  \
0                                             0  ...   
2                                             0  ...   
3                                             0  ...   
4                                             0  ...   
5                                             0  ...   

   Suspicious Intent Filter: USER_PRESENT  Hidden Apk  \
0                                       0           0   
2                                       0           0   
3                                       0           0   
4                                       0           0   
5                                       0           0   

   Sends SMS to Suspicious Number(s)  Package Domain Exists  \
0                                  0                      1   
2                                  0                      1   
3                                  0                      1   
4                                  0                      1   
5                                  0                      1   

   Reads phone data at startup  Sends SMS at startup  \
0                            0                     0   
2                            0                     0   
3                            0                     0   
4                            0                     0   
5                            0                     0   

   Starts service at startup  Sends SMS when receiving SMS  \
0                          0                             0   
2        

In [99]:
df.select_dtypes(exclude=['int','float'])

Classes with Ad prefix Total number of packages  \
0                           0                        3   
2                           0                        3   
3                           1                       95   
4                           0                        2   
5                           0                        1   
...                       ...                      ...   
162978                      0                        2   
162979                      0                        6   
162980                      0                        6   
162981                      1                        9   
162982                      0                       14   

       Total number of classes Number of classes in main package  \
0                           13                                10   
2                           14                                 0   
3                         1406                                 0   
4                           13                                 0   
5                           10                                 0   
...                        ...                               ...   
162978                      52                                 0   
162979                      53                                 0   
162980                      53                                 0   
162981                     148                                 0   
162982                     248                                14   

       Average class size Size of apk Number of images Number of files  \
0             1776.846154     1474118               88             330   
2             2001.857143     1616255               56             289   
3            10916.800142     2497258                3            1437   
4             2017.384615     1942479              181             269   
5                    3019      498359              204             283   
...                   ...         ...              ...             ...   
162978       57225.019231     1154683                1             114   
162979        7713.679245     1194699              193             263   
162980        7714.320755     1017805              191             261   
162981        7578.695946     3003014               97             266   
162982        7377.487903     1117891              259             606   

       Number of permissions Number of activities Number of services  \
0                          0                    1                  0   
2                          9                    1                  0   
3                          9                    7                  1   
4                          8                    1                  0   
5                          3                    1                  0   
...                      ...                  ...                ...   
162978                     3                    1                  0   
162979                     3                    1                  0   
162980                     0                    1                  0   
162981                    18                    5                  0   
162982                     4                   13                  0   

       Number of receivers  
0                        0  
2                        0  
3                        4  
4                        0  
5                        0  
...                    ...  
162978                   1  
162979                   0  
162980                   0  
162981                   0  
162982                   3  

[96732 rows x 12 columns]

In [100]:
cast_to_numeric_inplace(df)
df.select_dtypes(exclude=['int','float'])

Empty DataFrame
Columns: []
Index: [0, 2, 3, 4, 5, 6, 7, 8, 11, 12, 15, 16, 17, 19, 20, 21, 22, 26, 27, 29, 30, 32, 33, 37, 38, 39, 41, 42, 43, 44, 46, 47, 48, 49, 50, 52, 53, 55, 56, 57, 58, 60, 61, 62, 64, 67, 68, 72, 73, 74, 75, 76, 78, 80, 81, 82, 93, 94, 97, 98, 100, 102, 104, 105, 106, 107, 108, 109, 110, 111, 112, 114, 115, 116, 117, 122, 126, 127, 128, 129, 130, 131, 132, 133, 134, 137, 138, 140, 143, 144, 145, 146, 149, 152, 153, 154, 155, 156, 161, 162, ...]

[96732 rows x 0 columns]

In [101]:
df.head()

Dangerous Permission: ACCESS_SUPERUSER  \
0                                       0   
2                                       0   
3                                       0   
4                                       0   
5                                       0   

   Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE  \
0                                                  0      
2                                                  0      
3                                                  0      
4                                                  0      
5                                                  0      

   Dangerous Permission: CLEAR_APP_USER_DATA  \
0                                          0   
2                                          0   
3                                          0   
4                                          0   
5                                          0   

   Dangerous Permission: DELETE_CACHE_FILES  \
0                                         0   
2                                         0   
3                                         0   
4                                         0   
5                                         0   

   Dangerous Permission: DELETE_PACKAGES  \
0                                      0   
2                                      0   
3                                      0   
4                                      0   
5                                      0   

   Dangerous Permission: DISABLE_KEYGUARD  Dangerous Permission: FACTORY_TEST  \
0                                       0                                   0   
2                                       0                                   0   
3                                       0                                   0   
4                                       0                                   0   
5                                       0                                   0   

   Dangerous Permission: INSTALL_PACKAGES  \
0                                       0   
2                                       0   
3                                       0   
4                                       0   
5                                       0   

   Dangerous Permission: INJECT_EVENTS  \
0                                    0   
2                                    0   
3                                    0   
4                                    0   
5                                    0   

   Dangerous Permission: INTERNAL_SYSTEM_WINDOW  ...  \
0                                             0  ...   
2                                             0  ...   
3                                             0  ...   
4                                             0  ...   
5                                             0  ...   

   Suspicious Intent Filter: USER_PRESENT  Hidden Apk  \
0                                       0           0   
2                                       0           0   
3                                       0           0   
4                                       0           0   
5                                       0           0   

   Sends SMS to Suspicious Number(s)  Package Domain Exists  \
0                                  0                      1   
2                                  0                      1   
3                                  0                      1   
4                                  0                      1   
5                                  0                      1   

   Reads phone data at startup  Sends SMS at startup  \
0                            0                     0   
2                            0                     0   
3                            0                     0   
4                            0                     0   
5                            0                     0   

   Starts service at startup  Sends SMS when receiving SMS  \
0                          0                             0   
2        

In [102]:
df['class'].value_counts()

0    86562
1    10170
Name: class, dtype: int64

In [103]:
df.to_csv('Androcrawl_LIMPO.csv')

Aqui vou gerar uma versão do dataset limpo do Androidcrawl contendo apenas features catalogadas pela Tainá

In [104]:
df3 = pd.read_csv('datasets/AndroCrawl_CPI.csv', index_col=0, low_memory=False)
cols_to_keep = []
for col in df3.columns:
    if(col.replace("'", "") in df.columns):
        cols_to_keep.append(col.replace("'", ""))
df[cols_to_keep + ['class']].to_csv('Androcrawl_LIMPO_apiCalls_intents_permissions.csv')

## Android Permissions Cleaning Dataset

In [4]:
df2 = pd.read_csv('./Android_permissions_Cleaning.csv', index_col=0)
df2

App  \
0             Canada Post Corporation   
1                           Word Farm   
2                Fortunes of War FREE   
3      Better Keyboard: Avatar Purple   
4                                 NaN   
...                               ...   
29994                 Beechwood Hotel   
29995                             NaN   
29996                             NaN   
29997                             NaN   
29998                             NaN   

                                        Package          Category  \
0                        com.canadapost.android          Business   
1                     com.realcasualgames.words    Brain & Puzzle   
2                            fortunesofwar.free    Cards & Casino   
3      com.cc.betterkeyboard.skins.avatarpurple  Libraries & Demo   
4                                           NaN               NaN   
...                                         ...               ...   
29994                       smartstay.beechwood    Travel & Local   
29995                                       NaN               NaN   
29996                                       NaN               NaN   
29997                                       NaN               NaN   
29998                                       NaN               NaN   

                                             Description  Rating  \
0      Canada Post Mobile App gives you access to som...     3.1   
1      Speed and strategy combine in this exciting wo...     4.3   
2      Fortunes of War is a fast-paced, easy to learn...     4.1   
3      Skin for Better Keyboard featuring a glossy fe...     3.6   
4                                                    NaN     NaN   
...                                                  ...     ...   
29994  Beechwood is an Android hotel app for guests, ...     5.0   
29995                                                NaN     NaN   
29996                                                NaN     NaN   
29997                                                NaN     NaN   
29998                                                NaN     NaN   

       Number of ratings  Price  \
0                   77.0    0.0   
1                  199.0    0.0   
2                  243.0    0.0   
3                 2589.0    0.0   
4                    NaN    NaN   
...                  ...    ...   
29994                2.0    0.0   
29995                NaN    NaN   
29996                NaN    NaN   
29997                NaN    NaN   
29998                NaN    NaN   

                                            Related apps  \
0      {com.adaffix.pub.ca.android, com.kevinquan.gas...   
1      {air.com.zubawing.FastWordLite, com.joybits.do...   
2      {com.kevinquan.condado, hu.monsta.pazaak, net....   
3      {eu.gdumoulin.betterandroid.skins.transparent....   
4                                                    NaN   
...                                                  ...   
29994  {com.bluewaterpages, com.travelconnection.luxu...   
29995                                                NaN   
29996                                                NaN   
29997                                                NaN   
29998                                                NaN   

       Dangerous permissions count  Safe permissions count  ...  \
0                              7.0                     1.0  ...   
1                              NaN                     2.0  ...   
2                              1.0                     1.0  ...   
3                              0.0                     0.0  ...   
4                              NaN                     NaN  ...   
...                            ...                     ...  ...   
29994                          NaN                     1.0  ...   
29995                          NaN                     NaN  ...   
29996                          NaN                     NaN  ...   
29997                          NaN                     NaN  ...   
29998           

In [106]:
for col, unique_values in get_unique_values(df2):
    print(col, unique_values)

App ['Canada Post Corporation' 'Word Farm' 'Fortunes of War FREE' ...
 'Paid Taking Surveys' 'FORUM CU Mobile DeposZip' 'Beechwood Hotel']
Package ['com.canadapost.android' 'com.realcasualgames.words' 'fortunesofwar.free'
 ... 'com.v1_4.B06F370884C1473445134E7C.com'
 'com.Vertifi.DeposZip.P274074037' 'smartstay.beechwood']
Category ['Business' 'Brain & Puzzle' 'Cards & Casino' 'Libraries & Demo' nan
 'Personalization' 'Arcade & Action' 'Social' 'Health & Fitness'
 'Entertainment' 'Casual' 'Media & Video' 'Finance' 'Music & Audio'
 'Books & Reference' 'Tools' 'Education' 'Shopping' 'Travel & Local'
 'Lifestyle' 'Sports' 'Racing' 'Productivity' 'Communication'
 'Transportation' 'Comics' 'Medical' 'Sports Games' 'Photography'
 'News & Magazines' 'Weather']
Description ['Canada Post Mobile App gives you access to some of the the most popular tools on <a href="http://www.google.com/url?q=http://canadapost.ca&usg=AFQjCNGLsUtAzxPlFZBfYYGjv6mnRiMZJw" target="_blank">canadapost.ca</a>.  You can

Network communication : view network state (S) [ 1.  0. nan]
Phone calls : intercept outgoing calls (D) [ 0. nan  1.]
Phone calls : modify phone state (S) [ 0. nan  1.]
Phone calls : read phone state and identity (D) [ 0.  1. nan]
Services that cost you money : directly call phone numbers (D) [ 1.  0. nan]
Services that cost you money : send SMS messages (D) [ 0. nan  1.]
Storage : modify/delete USB storage contents modify/delete SD card contents (D) [ 1.  0. nan]
System tools : allow Wi-Fi Multicast reception (D) [ 0. nan  1.]
System tools : automatically start at boot (S) [ 0. nan  1.]
System tools : bluetooth administration (D) [ 0. nan  1.]
System tools : change Wi-Fi state (D) [ 0. nan  1.]
System tools : change background data usage setting (S) [ 0. nan]
System tools : change network connectivity (D) [ 0. nan  1.]
System tools : change your UI settings (D) [ 0. nan  1.]
System tools : delete all application cache data (D) [ 0. nan  1.]
System tools : disable keylock (D) [ 0. nan 

### 1) Verificar e remover valores faltantes

In [107]:
df2 = df2.dropna(subset=['Class'])

Existem amostras com valores faltantes (`NaN`) nas colunas `Dangerous permissions count` e `Safe permissions count`. O motivo disso é que há amostras de permissões com valores faltantes também. Logo, precisamos remover colunas (e não linhas/amostras, mais sobre isso a seguir) com features de permissões que tenham valores faltantes e recalcular `Dangerous permissions count` e `Safe permissions count`. 

Se tentássemos remover as linhas com permissões faltantes, nenhuma linha restaria:

In [108]:
get_danger_permission_cols = lambda : [col for col in df2.columns if '(D)' == col.replace(' ', '')[-3:]]
get_safe_permission_cols = lambda : [col for col in df2.columns if '(S)' == col.replace(' ', '')[-3:]]
permission_cols = get_danger_permission_cols() + get_safe_permission_cols()
df2.dropna(subset=permission_cols)

Empty DataFrame
Columns: [App, Package, Category, Description, Rating, Number of ratings, Price, Related apps, Dangerous permissions count, Safe permissions count, Default : Access DRM content. (S), Default : Access Email provider data (S), Default : Access all system downloads (S), Default : Access download manager. (S), Default : Advanced download manager functions. (S), Default : Audio File Access (S), Default : Install DRM content. (S), Default : Modify Google service configuration (S), Default : Modify Google settings (S), Default : Move application resources (S), Default : Read Google settings (S), Default : Send download notifications. (S), Default : Voice Search Shortcuts (S), Default : access SurfaceFlinger (S), Default : access checkin properties (S), Default : access the cache filesystem (S), Default : access to passwords for Google accounts (S), Default : act as an account authenticator (S), Default : bind to a wallpaper (S), Default : bind to an input method (S), Default : change screen orientation (S), Default : coarse (network-based) location (S), Default : control location update notifications (S), Default : control system backup and restore (S), Default : delete applications (S), Default : delete other applications' caches (S), Default : delete other applications' data (S), Default : directly call any phone numbers (S), Default : directly install applications (S), Default : disable or modify status bar (S), Default : discover known accounts (S), Default : display unauthorized windows (S), Default : enable or disable application components (S), Default : force application to close (S), Default : force device reboot (S), Default : full Internet access (S), Default : interact with a device admin (S), Default : manage application tokens (S), Default : mock location sources for testing (S), Default : modify battery statistics (S), Default : modify secure system settings (S), Default : modify the Google services map (S), Default : modify/delete USB storage contents modify/delete SD card contents (S), Default : monitor and control all application launching (S), Default : partial shutdown (S), Default : permanently disable device (S), Default : permission to install a location provider (S), Default : power device on or off (S), Default : press keys and control buttons (S), Default : prevent app switches (S), Default : read frame buffer (S), Default : read instant messages (S), Default : read phone state and identity (S), Default : record what you type and actions you take (S), Default : reset system to factory defaults (S), Default : run in factory test mode (S), Default : set time (S), Default : set wallpaper size hints (S), Default : start IM service (S), Default : update component usage statistics (S), Default : write contact data (S), Default : write instant messages (S), Development tools : enable application debugging (D), Development tools : limit number of running processes (D), Development tools : make all background applications close (D), Development tools : send Linux signals to applications (D), Hardware controls : change your audio settings (D), Hardware controls : control flashlight (S), Hardware controls : control vibrator (S), Hardware controls : record audio (D), Hardware controls : take pictures and videos (D), Hardware controls : test hardware (S), Network communication : Broadcast data messages to applications. (S), Network communication : control Near Field Communication (D), Network communication : create Bluetooth connections (D), Network communication : download files without notification (S), Network communication : full Internet access (D), Network communication : make/receive Internet calls (D), Network communication : receive data from Internet (S), Network communication : view Wi-Fi state (S), Network communication : view network state (S), Phone calls : intercept outgoing calls (D), Phone calls : modify phone state (S), Phone calls : read phone state and identity (D), Services that cost 

In [109]:
valid_cols = df2[permission_cols].dropna(axis=1).columns
df2 = df2.drop(columns=[col for col in permission_cols if col not in valid_cols])
df2

App  \
0             Canada Post Corporation   
1                           Word Farm   
2                Fortunes of War FREE   
3      Better Keyboard: Avatar Purple   
5             Ms Claus Live Wallpaper   
...                               ...   
29990                        EonPhone   
29991             Paid Taking Surveys   
29992           TeleNav GPS Navigator   
29993        FORUM CU Mobile DeposZip   
29994                 Beechwood Hotel   

                                        Package          Category  \
0                        com.canadapost.android          Business   
1                     com.realcasualgames.words    Brain & Puzzle   
2                            fortunesofwar.free    Cards & Casino   
3      com.cc.betterkeyboard.skins.avatarpurple  Libraries & Demo   
5               tmc.christmaslady.livewallpaper   Personalization   
...                                         ...               ...   
29990                          com.eonphone.sip     Communication   
29991     com.v1_4.B06F370884C1473445134E7C.com          Business   
29992            com.telenav.app.android.sprint    Travel & Local   
29993           com.Vertifi.DeposZip.P274074037           Finance   
29994                       smartstay.beechwood    Travel & Local   

                                             Description  Rating  \
0      Canada Post Mobile App gives you access to som...     3.1   
1      Speed and strategy combine in this exciting wo...     4.3   
2      Fortunes of War is a fast-paced, easy to learn...     4.1   
3      Skin for Better Keyboard featuring a glossy fe...     3.6   
5      Ms Claus Live Wallpaper<p>Find more Free apps ...     4.5   
...                                                  ...     ...   
29990  Eon Phone clearest VOIP calls to the USA and C...     3.0   
29991  Get paid for your opinion by taking easy onlin...     0.0   
29992  Get voice-guided GPS, maps, business listings ...     4.1   
29993  Make check deposits to your FORUM Credit Union...     4.8   
29994  Beechwood is an Android hotel app for guests, ...     5.0   

       Number of ratings  Price  \
0                   77.0    0.0   
1                  199.0    0.0   
2                  243.0    0.0   
3                 2589.0    0.0   
5                   16.0    0.0   
...                  ...    ...   
29990               47.0    0.0   
29991                0.0    0.0   
29992             2664.0    0.0   
29993                4.0    0.0   
29994                2.0    0.0   

                                            Related apps  \
0      {com.adaffix.pub.ca.android, com.kevinquan.gas...   
1      {air.com.zubawing.FastWordLite, com.joybits.do...   
2      {com.kevinquan.condado, hu.monsta.pazaak, net....   
3      {eu.gdumoulin.betterandroid.skins.transparent....   
5      {tmc.christmassanta.livewallpaper, tmc.winterh...   
...                                                  ...   
29990  {com.mytunudialer.activity, cz.acrobits.softph...   
29991  {es.feliperipoll.android.inga, com.appmakr.app...   
29992  {com.rocketouch.CRCKosher, com.cb.sfl, com.kos...   
29993  {com.forumcu.cuonline, com.msi.bktest, com.qua...   
29994  {com.bluewaterpages, com.travelconnection.luxu...   

       Dangerous permissions count  Safe permissions count  ...  \
0                              7.0                     1.0  ...   
1                              NaN                     2.0  ...   
2                              1.0                     1.0  ...   
3                              0.0                     0.0  ...   
5                              NaN                     1.0  ...   
...                            ...                     ...  ...   
29990                         11.0                     1.0  ...   
29991                          NaN                     1.0  ...   
29992                         15.0                     7.0  ...   
29993                          2.0                     0.0  ...   
29994           

In [110]:
# recalcular as colunas que contam permissões perigosas e seguras
df2['Dangerous permissions count'] = df2[get_danger_permission_cols()].sum(axis=1)
df2['Safe permissions count'] = df2[get_safe_permission_cols()].sum(axis=1)
df2

App  \
0             Canada Post Corporation   
1                           Word Farm   
2                Fortunes of War FREE   
3      Better Keyboard: Avatar Purple   
5             Ms Claus Live Wallpaper   
...                               ...   
29990                        EonPhone   
29991             Paid Taking Surveys   
29992           TeleNav GPS Navigator   
29993        FORUM CU Mobile DeposZip   
29994                 Beechwood Hotel   

                                        Package          Category  \
0                        com.canadapost.android          Business   
1                     com.realcasualgames.words    Brain & Puzzle   
2                            fortunesofwar.free    Cards & Casino   
3      com.cc.betterkeyboard.skins.avatarpurple  Libraries & Demo   
5               tmc.christmaslady.livewallpaper   Personalization   
...                                         ...               ...   
29990                          com.eonphone.sip     Communication   
29991     com.v1_4.B06F370884C1473445134E7C.com          Business   
29992            com.telenav.app.android.sprint    Travel & Local   
29993           com.Vertifi.DeposZip.P274074037           Finance   
29994                       smartstay.beechwood    Travel & Local   

                                             Description  Rating  \
0      Canada Post Mobile App gives you access to som...     3.1   
1      Speed and strategy combine in this exciting wo...     4.3   
2      Fortunes of War is a fast-paced, easy to learn...     4.1   
3      Skin for Better Keyboard featuring a glossy fe...     3.6   
5      Ms Claus Live Wallpaper<p>Find more Free apps ...     4.5   
...                                                  ...     ...   
29990  Eon Phone clearest VOIP calls to the USA and C...     3.0   
29991  Get paid for your opinion by taking easy onlin...     0.0   
29992  Get voice-guided GPS, maps, business listings ...     4.1   
29993  Make check deposits to your FORUM Credit Union...     4.8   
29994  Beechwood is an Android hotel app for guests, ...     5.0   

       Number of ratings  Price  \
0                   77.0    0.0   
1                  199.0    0.0   
2                  243.0    0.0   
3                 2589.0    0.0   
5                   16.0    0.0   
...                  ...    ...   
29990               47.0    0.0   
29991                0.0    0.0   
29992             2664.0    0.0   
29993                4.0    0.0   
29994                2.0    0.0   

                                            Related apps  \
0      {com.adaffix.pub.ca.android, com.kevinquan.gas...   
1      {air.com.zubawing.FastWordLite, com.joybits.do...   
2      {com.kevinquan.condado, hu.monsta.pazaak, net....   
3      {eu.gdumoulin.betterandroid.skins.transparent....   
5      {tmc.christmassanta.livewallpaper, tmc.winterh...   
...                                                  ...   
29990  {com.mytunudialer.activity, cz.acrobits.softph...   
29991  {es.feliperipoll.android.inga, com.appmakr.app...   
29992  {com.rocketouch.CRCKosher, com.cb.sfl, com.kos...   
29993  {com.forumcu.cuonline, com.msi.bktest, com.qua...   
29994  {com.bluewaterpages, com.travelconnection.luxu...   

       Dangerous permissions count  Safe permissions count  ...  \
0                              6.0                     0.0  ...   
1                              3.0                     1.0  ...   
2                              1.0                     0.0  ...   
3                              0.0                     0.0  ...   
5                              3.0                     0.0  ...   
...                            ...                     ...  ...   
29990                         11.0                     1.0  ...   
29991                          3.0                     0.0  ...   
29992                         15.0                     6.0  ...   
29993                          2.0                     0.0  ...   
29994           

In [111]:
print("Colunas que ainda possuem NaN's:")
for col, unique_values in get_unique_values(df2):
    if (any([pd.isna(value) for value in unique_values])):
        print(col)

Colunas que ainda possuem NaN's:
App
Description
Related apps


### 2) Remoção de amostras com valores duplicados nas colunas

In [112]:
df2 = df2[~df2.duplicated()]
df2

App  \
0             Canada Post Corporation   
1                           Word Farm   
2                Fortunes of War FREE   
3      Better Keyboard: Avatar Purple   
5             Ms Claus Live Wallpaper   
...                               ...   
29990                        EonPhone   
29991             Paid Taking Surveys   
29992           TeleNav GPS Navigator   
29993        FORUM CU Mobile DeposZip   
29994                 Beechwood Hotel   

                                        Package          Category  \
0                        com.canadapost.android          Business   
1                     com.realcasualgames.words    Brain & Puzzle   
2                            fortunesofwar.free    Cards & Casino   
3      com.cc.betterkeyboard.skins.avatarpurple  Libraries & Demo   
5               tmc.christmaslady.livewallpaper   Personalization   
...                                         ...               ...   
29990                          com.eonphone.sip     Communication   
29991     com.v1_4.B06F370884C1473445134E7C.com          Business   
29992            com.telenav.app.android.sprint    Travel & Local   
29993           com.Vertifi.DeposZip.P274074037           Finance   
29994                       smartstay.beechwood    Travel & Local   

                                             Description  Rating  \
0      Canada Post Mobile App gives you access to som...     3.1   
1      Speed and strategy combine in this exciting wo...     4.3   
2      Fortunes of War is a fast-paced, easy to learn...     4.1   
3      Skin for Better Keyboard featuring a glossy fe...     3.6   
5      Ms Claus Live Wallpaper<p>Find more Free apps ...     4.5   
...                                                  ...     ...   
29990  Eon Phone clearest VOIP calls to the USA and C...     3.0   
29991  Get paid for your opinion by taking easy onlin...     0.0   
29992  Get voice-guided GPS, maps, business listings ...     4.1   
29993  Make check deposits to your FORUM Credit Union...     4.8   
29994  Beechwood is an Android hotel app for guests, ...     5.0   

       Number of ratings  Price  \
0                   77.0    0.0   
1                  199.0    0.0   
2                  243.0    0.0   
3                 2589.0    0.0   
5                   16.0    0.0   
...                  ...    ...   
29990               47.0    0.0   
29991                0.0    0.0   
29992             2664.0    0.0   
29993                4.0    0.0   
29994                2.0    0.0   

                                            Related apps  \
0      {com.adaffix.pub.ca.android, com.kevinquan.gas...   
1      {air.com.zubawing.FastWordLite, com.joybits.do...   
2      {com.kevinquan.condado, hu.monsta.pazaak, net....   
3      {eu.gdumoulin.betterandroid.skins.transparent....   
5      {tmc.christmassanta.livewallpaper, tmc.winterh...   
...                                                  ...   
29990  {com.mytunudialer.activity, cz.acrobits.softph...   
29991  {es.feliperipoll.android.inga, com.appmakr.app...   
29992  {com.rocketouch.CRCKosher, com.cb.sfl, com.kos...   
29993  {com.forumcu.cuonline, com.msi.bktest, com.qua...   
29994  {com.bluewaterpages, com.travelconnection.luxu...   

       Dangerous permissions count  Safe permissions count  ...  \
0                              6.0                     0.0  ...   
1                              3.0                     1.0  ...   
2                              1.0                     0.0  ...   
3                              0.0                     0.0  ...   
5                              3.0                     0.0  ...   
...                            ...                     ...  ...   
29990                         11.0                     1.0  ...   
29991                          3.0                     0.0  ...   
29992                         15.0                     6.0  ...   
29993                          2.0                     0.0  ...   
29994           

### 3) Remoção de características irrelevantes

In [113]:
df2 = df2.drop(columns=['App', 'Package', 'Category','Description', 'Related apps', 'Price'])
df2 = drop_irrelevant_columns(df2)
df2

21 colunas irrelevantes:
['Default : Modify Google service configuration (S)', 'Default : access to passwords for Google accounts (S)', 'Default : act as an account authenticator (S)', 'Default : coarse (network-based) location (S)', 'Default : discover known accounts (S)', 'Default : full Internet access (S)', 'Default : mock location sources for testing (S)', 'Default : modify/delete USB storage contents modify/delete SD card contents (S)', 'Default : permanently disable device (S)', 'Default : read instant messages (S)', 'Default : reset system to factory defaults (S)', 'Default : run in factory test mode (S)', 'Default : set wallpaper size hints (S)', 'Default : start IM service (S)', 'Default : write contact data (S)', 'Default : write instant messages (S)', 'Network communication : download files without notification (S)', 'System tools : change background data usage setting (S)', 'Your accounts : Blogger (D)', 'Your messages : read Gmail attachment previews (D)', 'Your messages 

Rating  Number of ratings  Dangerous permissions count  \
0         3.1               77.0                          6.0   
1         4.3              199.0                          3.0   
2         4.1              243.0                          1.0   
3         3.6             2589.0                          0.0   
5         4.5               16.0                          3.0   
...       ...                ...                          ...   
29990     3.0               47.0                         11.0   
29991     0.0                0.0                          3.0   
29992     4.1             2664.0                         15.0   
29993     4.8                4.0                          2.0   
29994     5.0                2.0                          3.0   

       Safe permissions count  Default : Access Email provider data (S)  \
0                         0.0                                       0.0   
1                         1.0                                       0.0   
2                         0.0                                       0.0   
3                         0.0                                       0.0   
5                         0.0                                       0.0   
...                       ...                                       ...   
29990                     1.0                                       0.0   
29991                     0.0                                       0.0   
29992                     6.0                                       0.0   
29993                     0.0                                       0.0   
29994                     1.0                                       0.0   

       Default : Advanced download manager functions. (S)  \
0                                                    0.0    
1                                                    0.0    
2                                                    0.0    
3                                                    0.0    
5                                                    0.0    
...                                                  ...    
29990                                                0.0    
29991                                                0.0    
29992                                                0.0    
29993                                                0.0    
29994                                                0.0    

       Default : Audio File Access (S)  Default : Install DRM content. (S)  \
0                                  0.0                                 0.0   
1                                  0.0                                 0.0   
2                                  0.0                                 0.0   
3                                  0.0                                 0.0   
5                                  0.0                                 0.0   
...                                ...                                 ...   
29990                              0.0                                 0.0   
29991                              0.0                                 0.0   
29992                              0.0                                 0.0   
29993                              0.0                                 0.0   
29994                              0.0                                 0.0   

       Default : Modify Google settings (S)  \
0                                       0.0   
1                                       0.0   
2                                       0.0   
3                                       0.0   
5                                       0.0   
...                                     ...   
29990                                   0.0   
29991                                   0.0   
29992                                   0.0   
29993                                   0.0   
29994                                   0.0   

       Default : Move application resources (S)  ...  \
0                                 

### 4) Conversão de tipos de dados

Apenas a coluna `Class` precisa ser corrigida

In [114]:
print(df2.select_dtypes(exclude=['int', 'float']))
print(df2['Class'].unique())
df2['Class'] = df2['Class'].replace({'Benign' : 0, 'Malware' : 1})

         Class
0       Benign
1       Benign
2       Benign
3       Benign
5       Benign
...        ...
29990  Malware
29991   Benign
29992  Malware
29993  Malware
29994  Malware

[27298 rows x 1 columns]
['Benign' 'Malware']


In [115]:
df2

Rating  Number of ratings  Dangerous permissions count  \
0         3.1               77.0                          6.0   
1         4.3              199.0                          3.0   
2         4.1              243.0                          1.0   
3         3.6             2589.0                          0.0   
5         4.5               16.0                          3.0   
...       ...                ...                          ...   
29990     3.0               47.0                         11.0   
29991     0.0                0.0                          3.0   
29992     4.1             2664.0                         15.0   
29993     4.8                4.0                          2.0   
29994     5.0                2.0                          3.0   

       Safe permissions count  Default : Access Email provider data (S)  \
0                         0.0                                       0.0   
1                         1.0                                       0.0   
2                         0.0                                       0.0   
3                         0.0                                       0.0   
5                         0.0                                       0.0   
...                       ...                                       ...   
29990                     1.0                                       0.0   
29991                     0.0                                       0.0   
29992                     6.0                                       0.0   
29993                     0.0                                       0.0   
29994                     1.0                                       0.0   

       Default : Advanced download manager functions. (S)  \
0                                                    0.0    
1                                                    0.0    
2                                                    0.0    
3                                                    0.0    
5                                                    0.0    
...                                                  ...    
29990                                                0.0    
29991                                                0.0    
29992                                                0.0    
29993                                                0.0    
29994                                                0.0    

       Default : Audio File Access (S)  Default : Install DRM content. (S)  \
0                                  0.0                                 0.0   
1                                  0.0                                 0.0   
2                                  0.0                                 0.0   
3                                  0.0                                 0.0   
5                                  0.0                                 0.0   
...                                ...                                 ...   
29990                              0.0                                 0.0   
29991                              0.0                                 0.0   
29992                              0.0                                 0.0   
29993                              0.0                                 0.0   
29994                              0.0                                 0.0   

       Default : Modify Google settings (S)  \
0                                       0.0   
1                                       0.0   
2                                       0.0   
3                                       0.0   
5                                       0.0   
...                                     ...   
29990                                   0.0   
29991                                   0.0   
29992                                   0.0   
29993                                   0.0   
29994                                   0.0   

       Default : Move application resources (S)  ...  \
0                                 

In [116]:
df2['Class'].value_counts()

1    18224
0     9074
Name: Class, dtype: int64

In [117]:
df2.to_csv("Android_Permissions_Cleaning_Dataset_LIMPO.csv")